File: sol/2104.py
class Solution:
    def subArrayRanges(self, nums: List[int]) -> int:
        n = len(nums)
        memo = [0] * n
        memo[0] = nums[0]
        st = [0]
        res = 0
        for i in range(1, n):
            while len(st) != 0 and nums[st[-1]] < nums[i]:
                st.pop()
            if len(st) == 0:
                memo[i] = (i + 1) * nums[i]
            else:
                j = st[-1]
                memo[i] = memo[j] + (i - j) * nums[i]
            st.append(i)
            res += memo[i]
        st = [0]
        for i in range(1, n):
            while len(st) != 0 and nums[st[-1]] > nums[i]:
                st.pop()
            if len(st) == 0:
                memo[i] = (i + 1) * nums[i]
            else:
                j = st[-1]
                memo[i] = memo[j] + (i - j) * nums[i]
            st.append(i)
            res -= memo[i]
        return res

File: sol/2104/2104.cpp
struct Solution {
    using ll = long long;
    long long subArrayRanges(vector<int>& nums) {
        int n = nums.size();
        vector<ll> memo(n, 0);
        memo[0] = nums[0];
        vector<int> st(1, 0);
        ll res = 0;
        for (int i = 1; i < n; ++i){
            while (!st.empty() && nums[st.back()] < nums[i]){
                st.pop_back();
            }
            if (st.empty()){
                memo[i] = static_cast<ll>(i + 1) * nums[i];
            } else {
                int j = st.back();
                memo[i] = memo[j] + static_cast<ll>(i - j) * nums[i];
            }
            st.push_back(i);
            res += memo[i];
        }
        st.clear();
        st.push_back(0);
        for (int i = 1; i < n; ++i){
            while (!st.empty() && nums[st.back()] > nums[i]){
                st.pop_back();
            }
            if (st.empty()){
                memo[i] = static_cast<ll>(i + 1) * nums[i];
            } else {
                int j = st.back();
                memo[i] = memo[j] + static_cast<ll>(i - j) * nums[i];
            }
            st.push_back(i);
            res -= memo[i];
        }
        return res;
    }
};
